### PR TITLE
preliminary and incomplete implementation of java pec methods: InstantiateParticle and DefineHandle

### PR DESCRIPTION
**Pull request by mmandlis**

### CODE REVIEW

File: src/javaharness/java/arcs/api/ParticleExecutionContextImpl.java [hunk 1]
```diff
@@ -1,22 +1,52 @@
...
     }
 
     @Override
-    public NativeParticle instantiateParticle(String name) {
-        return particleLoader.loadParticle(name).flatMap(x -> Optional.of(x.createParticle()))
-                .orElse(null);
+    public NativeParticle instantiateParticle(ParticleSpec spec, Map<String, StorageProxy> proxies) {
+        // TODO: use the full spec.implPath, instead of the filename.
+        NativeParticle particle = particleLoader.loadParticle(spec.getFileName()).flatMap(
+            x -> Optional.of(x.createParticle())).orElse(null);
```
Comment (cromwellian) [inline src/javaharness/java/arcs/api/ParticleExecutionContextImpl.java]: Optional.ofNullable()

File: src/javaharness/java/arcs/api/Thing.java [hunk 2]
```diff
@@ -18,4 +20,12 @@ public boolean equals(Object other) {
     public int hashCode() {
         return get().hashCode();
     }
+
+    public NativeParticle getParticle() {
+        return this.thing instanceof NativeParticle ? Optional.of((NativeParticle)this.thing).orElse(null) : null;
```
Comment (cromwellian) [inline src/javaharness/java/arcs/api/Thing.java]: space after paren
(NativeParticle) this

File: src/javaharness/java/arcs/api/Thing.java [hunk 3]
```diff
@@ -1,5 +1,7 @@
 package arcs.api;
 
+import java.util.Optional;
+
 public class Thing<T> {
     public final T thing;
```
Comment (cromwellian) [inline src/javaharness/java/arcs/api/Thing.java]: add an assertion in constructor that this is never null

File: src/javaharness/java/arcs/api/Thing.java [hunk 4]
```diff
@@ -18,4 +20,12 @@ public boolean equals(Object other) {
     public int hashCode() {
         return get().hashCode();
     }
+
+    public NativeParticle getParticle() {
```
Comment (cromwellian) [inline src/javaharness/java/arcs/api/Thing.java]: assert is a particle

**Comment by mmandlis** [comment]
Addressed comments, PTAL

File: src/javaharness/java/arcs/api/Thing.java (hunk 2 above)
Comment (mmandlis) [reply src/javaharness/java/arcs/api/Thing.java]: done

File: src/javaharness/java/arcs/api/Thing.java (hunk 3 above)
Comment (mmandlis) [reply src/javaharness/java/arcs/api/Thing.java]: done

File: src/javaharness/java/arcs/api/Thing.java (hunk 4 above)
Comment (mmandlis) [reply src/javaharness/java/arcs/api/Thing.java]: done

File: src/javaharness/java/arcs/api/ParticleExecutionContextImpl.java (hunk 1 above)
Comment (mmandlis) [reply src/javaharness/java/arcs/api/ParticleExecutionContextImpl.java]: done.

**Comment by mmandlis** [comment]
Addressed comments, PTAL

**Pull request merged**